File: java/com/joshmanisdabomb/lcc/block/FunctionalTNTBlock.java
package com.joshmanisdabomb.lcc.block;

import com.joshmanisdabomb.lcc.misc.AdaptedFromSource;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.TNTBlock;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.item.TNTEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.projectile.AbstractArrowEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.Hand;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvents;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.BlockRayTraceResult;
import net.minecraft.world.Explosion;
import net.minecraft.world.World;

import javax.annotation.Nullable;

public class FunctionalTNTBlock extends TNTBlock {

    private final TNTFactory factory;

    public FunctionalTNTBlock(TNTFactory factory, Properties properties) {
        super(properties);
        this.factory = factory;
    }

    public FunctionalTNTBlock(TNTFactory factory, boolean unstable, Properties properties) {
        this(factory, properties);
        this.setDefaultState(this.stateContainer.getBaseState().with(UNSTABLE, unstable));
    }

    @Override
    @AdaptedFromSource
    public void onBlockAdded(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving) {
        if (oldState.getBlock() != state.getBlock()) {
            if (worldIn.isBlockPowered(pos)) {
                this.explode(worldIn, pos, null);
                worldIn.removeBlock(pos, false);
            }
        }
    }

    @Override
    @AdaptedFromSource
    public void neighborChanged(BlockState state, World worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving) {
        if (worldIn.isBlockPowered(pos)) {
            this.explode(worldIn, pos, null);
            worldIn.removeBlock(pos, false);
        }

    }

    /**
     * Called before the Block is set to air in the world. Called regardless of if the player's tool can actually collect
     * this block
     */
    @Override
    @AdaptedFromSource
    public void onBlockHarvested(World worldIn, BlockPos pos, BlockState state, PlayerEntity player) {
        if (!worldIn.isRemote() && !player.isCreative() && state.get(UNSTABLE)) {
            this.explode(worldIn, pos, null);
        }
        worldIn.playEvent(player, 2001, pos, getStateId(state));
    }

    /**
     * Called when this Block is destroyed by an Explosion
     */
    @Override
    @AdaptedFromSource
    public void onExplosionDestroy(World worldIn, BlockPos pos, Explosion explosionIn) {
        if (!worldIn.isRemote) {
            TNTEntity tntentity = factory.create(worldIn, (double)((float)pos.getX() + 0.5F), (double)pos.getY(), (double)((float)pos.getZ() + 0.5F), explosionIn.getExplosivePlacedBy());
            tntentity.setFuse((short)(worldIn.rand.nextInt(tntentity.getFuse() / 4) + tntentity.getFuse() / 8));
            worldIn.addEntity(tntentity);
        }
    }

    public void explode(World world, BlockPos pos, @Nullable LivingEntity entity) {
        if (!world.isRemote) {
            TNTEntity tntentity = factory.create(world, (double)((float)pos.getX() + 0.5F), (double)pos.getY(), (double)((float)pos.getZ() + 0.5F), entity);
            world.addEntity(tntentity);
            world.playSound(null, tntentity.posX, tntentity.posY, tntentity.posZ, SoundEvents.ENTITY_TNT_PRIMED, SoundCategory.BLOCKS, 1.0F, 1.0F);

        }
    }

    @Override
    @AdaptedFromSource
    public boolean onBlockActivated(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, Hand handIn, BlockRayTraceResult hit) {
        ItemStack itemstack = player.getHeldItem(handIn);
        Item item = itemstack.getItem();
        if (item != Items.FLINT_AND_STEEL && item != Items.FIRE_CHARGE) {
            return super.onBlockActivated(state, worldIn, pos, player, handIn, hit);
        } else {
            this.explode(worldIn, pos, player);
            worldIn.setBlockState(pos, Blocks.AIR.getDefaultState(), 11);
            if (item == Items.FLINT_AND_STEEL) {
                itemstack.damageItem(1, player, (p_220287_1_) -> {
                    p_220287_1_.sendBreakAnimation(handIn);
                });
            } else {
                itemstack.shrink(1);
            }

            return true;
        }
    }

    @Override
    @AdaptedFromSource
    public void onProjectileCollision(World worldIn, BlockState state, BlockRayTraceResult hit, Entity projectile) {
        if (!worldIn.isRemote && projectile instanceof AbstractArrowEntity) {
            AbstractArrowEntity abstractarrowentity = (AbstractArrowEntity)projectile;
            Entity entity = abstractarrowentity.getShooter();
            if (abstractarrowentity.isBurning()) {
                BlockPos blockpos = hit.getPos();
                this.explode(worldIn, blockpos, entity instanceof LivingEntity ? (LivingEntity)entity : null);
                worldIn.removeBlock(blockpos, false);
            }
        }
    }

    public interface TNTFactory {
        TNTEntity create(World world, double x, double y, double z, LivingEntity placer);
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/block/FunctionalTNTBlock.kt
package com.joshmanisdabomb.lcc.block

import com.joshmanisdabomb.lcc.extensions.isSurvival
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.block.TntBlock
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.TntEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.projectile.ProjectileEntity
import net.minecraft.item.Items
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.explosion.Explosion

class FunctionalTNTBlock(val factory: (world: World, x: Double, y: Double, z: Double, placer: LivingEntity?) -> TntEntity, settings: Settings, unstable: Boolean = false) : TntBlock(settings) {

    init {
        defaultState = stateManager.defaultState.with(UNSTABLE, unstable)
    }

    override fun onBlockAdded(state: BlockState, world: World, pos: BlockPos, oldState: BlockState, notify: Boolean) {
        if (!oldState.isOf(state.block)) {
            if (world.isReceivingRedstonePower(pos)) {
                ignite(world, pos, null)
                world.removeBlock(pos, false)
            }
        }
    }

    override fun neighborUpdate(state: BlockState, world: World, pos: BlockPos, block: Block, fromPos: BlockPos, notify: Boolean) {
        if (world.isReceivingRedstonePower(pos)) {
            ignite(world, pos, null)
            world.removeBlock(pos, false)
        }
    }

    override fun onBreak(world: World, pos: BlockPos, state: BlockState, player: PlayerEntity) {
        if (!world.isClient() && player.isSurvival && state[UNSTABLE]) {
            ignite(world, pos, null)
        }
        super.onBreak(world, pos, state.with(UNSTABLE, false), player)
    }

    override fun onDestroyedByExplosion(world: World, pos: BlockPos, explosion: Explosion) {
        if (!world.isClient) {
            val tntEntity = factory(world, pos.x.toDouble() + 0.5, pos.y.toDouble(), pos.z.toDouble() + 0.5, explosion.causingEntity)
            val i = tntEntity.fuse
            tntEntity.fuse = world.random.nextInt(i / 4) + i / 8
            world.spawnEntity(tntEntity)
        }
    }

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        val stack = player.getStackInHand(hand)
        return if (!stack.isOf(Items.FLINT_AND_STEEL) && !stack.isOf(Items.FIRE_CHARGE)) {
            super.onUse(state, world, pos, player, hand, hit)
        } else {
            ignite(world, pos, player)
            world.setBlockState(pos, Blocks.AIR.defaultState, 11)
            if (player.isSurvival) {
                if (stack.isOf(Items.FLINT_AND_STEEL)) {
                    stack.damage(1, player) { it.sendToolBreakStatus(hand) }
                } else {
                    stack.decrement(1)
                }
            }
            ActionResult.success(world.isClient)
        }
    }

    override fun onProjectileHit(world: World, state: BlockState, hit: BlockHitResult, projectile: ProjectileEntity) {
        if (!world.isClient) {
            val owner = projectile.owner
            if (projectile.isOnFire) {
                val blockPos = hit.blockPos
                ignite(world, blockPos, owner as? LivingEntity)
                world.removeBlock(blockPos, false)
            }
        }
    }

    fun ignite(world: World, pos: BlockPos, igniter: LivingEntity?) {
        if (!world.isClient) {
            val tntEntity = factory(world, pos.x.toDouble() + 0.5, pos.y.toDouble(), pos.z.toDouble() + 0.5, igniter)
            world.spawnEntity(tntEntity)
            world.playSound(null, tntEntity.x, tntEntity.y, tntEntity.z, SoundEvents.ENTITY_TNT_PRIMED, SoundCategory.BLOCKS, 1.0f, 1.0f)
        }
    }

}